packages/flutter/lib/src/cupertino/page_scaffold.dart

```diff
@@ -0,0 +1,72 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'colors.dart';
+import 'nav_bar.dart';
+
+/// Implements a single iOS application page's layout.
+///
+/// The scaffold lays out the navigation bar on top and the content between or
+/// behind the navigation bar.
+///
+/// See also:
+///
+///  * [CupertinoPageRoute] a modal page route that typically hosts a [CupertinoPageRoute]
+///    with support for iOS style page transitions.
+class CupertinoPageScaffold extends StatelessWidget {
+  const CupertinoPageScaffold({
+    Key key,
+    this.navigationBar,
+    @required this.child,
+  }) : assert(child != null),
+       super(key: key);
+
+  /// The [navigationBar], typically a [CupertinoNavigationBar], is drawn at the
+  /// top of the screen.
+  ///
+  /// If translucent, the main content may slide behind it.
+  /// Otherwise, the main content's top margin will be offset by its height.
+  // TODO(xster): document its page transition animation when ready
+  final PreferredSizeWidget navigationBar;
+
+  /// Widget to show in the main content area.
+  ///
+  /// Content can slide under the [navigationBar] when they're translucent.
+  final Widget child;
+
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> stacked = <Widget>[];
+    double topPadding = 0.0;
+    if (navigationBar != null) {
+      topPadding += navigationBar.preferredSize.height;
+      if (topPadding > 0.0)
+        topPadding += MediaQuery.of(context).padding.top;
+    }
+
+    // The main content being at the bottom is added to the stack first.
+    stacked.add(new Padding(
+      padding: new EdgeInsets.only(top: topPadding),
```

Comment:
if you apply the MediaQuery top padding here, you should make sure to introduce a child MediaQuery that removes that padding.

Reply:
I think if there's a padding here, I'd expect its descendants to never interact with window top paddings again kinda like material scaffold's body

Reply:
What if the descendant is a material Scaffold? or an AppBar? Or a PageScaffold?

Reply:
(it would be a factory constructor that grabs the ambient MediaQuery, and builds a new one that's identical except with the relevant padding(s) set to 0.0)

Reply:
Did you mean if there were a page scaffold with an app bar in the app bar position and another app bar in the body position? Ya, that would be problematic. 

Will make the change. 

Change to zero seems like a pretty specific API though. Why not just MediaQuery.of(context).copyWith?

Reply:
Actually my thing is a bit cumbersome, it'd have to kinda be
```
new MediaQuery(
  data: MediaQuery.of(context).data.copyWith(
    padding: MediaQuery.of(context).data.padding.copyWith(
      top: ...
```

Though @cbracken had a nice idea for a nice API that wraps all this stuff in a widget. I'll just make a MediaQuery the normal way and let Chris do the nicer API.

Reply:
The SafeArea widget that @cbracken and I were talking about would also introduce padding, and would need to do the removing of the padding in a secondary MediaQuery as well. So it's actually a superset of the functionality you need here, and therefore would be easily handled by the `new MediaQuery.removePadding(context, removeTop: true)`-style API. But whatever floats your boat. :-)

Reply:
Have to get on plane :S. Manually created a MediaQuery for now